### PR TITLE
Close #22494: Split theme transparency checkbox to column

### LINKED ISSUE #22494: Themes window: Add a tooltip to the checkbox explaining what they’re for
(Transparancy, if you didn’t know.)

### PR DESCRIPTION
**Pull request by lewyche**
image
image

Address: #22494

### CODE REVIEW

**Comment by AaronVanGeffen** [comment]
Nice work. Going by the screenshots, I see labels from the first column can bleed into the second column, though. I think this should be mitigated by replacing `DrawTextBasic` with `DrawTextEllipsised`.

To prevent label cutoff, I think we should consider increasing the window dimensions slightly. Alternatively, the 'Palette' column could be shrunk a little, but I think that will look unbalanced with the 'Transparency' column.

**Comment by Gymnasiast** [comment]
I think it might be better to just employ the "wrapped" text drawing function. Especially as other languages tend to result in longer strings.

**Comment by AaronVanGeffen** [comment]
> I think it is better to just employ the "wrapped" text drawing function. Especially as other languages tend to result in longer strings.

That would work for entries with more than one colour 'row', but would likely complicate things for single-row entries, like in the second screenshot 🤔

**Comment by Gymnasiast** [comment]
Would it, really? Isn’t it just a matter of saving the highest `y` coordinate?

**Comment by AaronVanGeffen** [comment]
You're right that it won't be a problem for drawing. It does complicate mouse events, though, as row heights would suddenly be affected by the label height. Similarly, getting the height of the scrollview would no longer be trivial.

**Comment by lewyche** [comment]
I've already delt with problems such as the mouse events and the height of the scroll view when I made the rows adapt to the number of colours, but making the text wrap and having an empty row might not look the prettiest.

**Comment by Gymnasiast** [comment]
I cannot accept lots of cut off titles in other languages, sorry. This game is not just made for people who play in English.

**Comment by AaronVanGeffen** [comment]
That is why I suggested the following before:

> To prevent label cutoff, I think we should consider increasing the window dimensions slightly. Alternatively, the 'Palette' column could be shrunk a little, but I think that will look unbalanced with the 'Transparency' column.

**Comment by lewyche** [comment]
I think extending the window so that it can fit most languages comfortably, and wrapping the text for languages with longer strings is a good comprise.

**Comment by Gymnasiast** [comment]
Ok, that would work for me.

**Comment by lewyche** [comment]
Testing in Russian:
wrap2
wrap1
I didn't end up increasing the window dimensions, since in most languages we don't need to make an empty row anyways.

File: data/language/en-GB.txt [hunk 1]
```diff
@@ -3740,6 +3740,7 @@ STR_6667    :Fauna
 STR_6668    :Staff cheats
 STR_6669    :Show staff cheats
 STR_6670    :Guest behaviour
+STR_6671    :Transparency
```
Comment (Gymnasiast) [inline data/language/en-GB.txt:3743]:
```suggestion
STR_6671    :Transparent
```

File: src/openrct2-ui/windows/Themes.cpp [hunk 2]
```diff
@@ -691,18 +703,27 @@ static WindowClass window_themes_tab_7_classes[] = {
...
             {
-                int32_t y2 = screenCoords.y % _row_height;
-                _colour_index_1 = screenCoords.y / _row_height;
-                _colour_index_2 = ((screenCoords.x - _button_offset_x) / 12);
+                _colour_index_1 = GetColourIndex(screenCoords.y);
 
                 WindowClass wc = GetWindowClassTabIndex(_colour_index_1);
                 int32_t numColours = ThemeDescGetNumColours(wc);
+
+                // add colour to account for text wrapping
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/Themes.cpp:713]:
```suggestion
                // If the window description wraps, make sure the row is high enough.
```

File: src/openrct2-ui/windows/Themes.cpp [hunk 3]
```diff
@@ -759,18 +781,33 @@ static WindowClass window_themes_tab_7_classes[] = {
             screenCoords.y = 0;
             for (int32_t i = 0; i < GetColourSchemeTabCount(); i++)
             {
+                WindowClass wc = GetWindowClassTabIndex(i);
+                int32_t numColours = ThemeDescGetNumColours(wc);
+
+                // make an empty row to account for text wrapping
+                bool empty_row = false;
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/Themes.cpp:788]: Change this to camelCase: `emptyRow`. This apply to any variable you specify.

File: src/openrct2-ui/windows/Themes.cpp (hunk 2 above)
Comment (Gymnasiast) [reply src/openrct2-ui/windows/Themes.cpp:713]: Also, I see this pattern multiple times in you changes. To me, it seems you could extract that out into a (static) function.

File: src/openrct2-ui/windows/Themes.cpp [hunk 4]
```diff
@@ -829,6 +873,56 @@ static WindowClass window_themes_tab_7_classes[] = {
             return classes[index];
         }
 
+        int8_t GetColourIndex(int32_t y)
+        {
+            int8_t total = 0;
```
Comment (Gymnasiast) [inline src/openrct2-ui/windows/Themes.cpp:878]: This seems to count the number of rows (occupied or not). In that case, `numRows` seems like a more appropriate name. This comment also applies to other cases where you count the number of rows.

**Review by Gymnasiast — CHANGES_REQUESTED** [review]
Looking through the code, I found a few things. Make sure to pay special attention to the line notes that apply to multiple places.

**Comment by Gymnasiast** [comment]
@lewyche Did you address all the points? If so, please say so and I’ll re-review.

**Comment by lewyche** [comment]
The PR should be ready to review now.

File: src/openrct2-ui/windows/Themes.cpp [hunk 5]
```diff
@@ -71,10 +73,12 @@ namespace OpenRCT2::Ui::Windows
     static constexpr int32_t WW = 320;
     static constexpr int32_t WH = 107;
 
+    const uint16_t _window_header_size = 152;
```
Comment (AaronVanGeffen) [inline src/openrct2-ui/windows/Themes.cpp:76]:
```suggestion
    const uint16_t kWindowHeaderWidth = 152;
```

**Comment by AaronVanGeffen** [comment]
The colour dropdowns are offset a bit strangely:
Screenshot 2024-09-11 at 09 49 03

Normally, we show dropdowns directly underneath the widget that called them, i.e. horizontally aligned. If you don't want to obstruct the elements underneath, I suggest we align them vertically with the start of the caller element.

**Comment by lewyche** [comment]
The dropdowns are now vertically aligned with the button, and changed _window_header_size to kWindowHeaderWidth
image

File: src/openrct2-ui/windows/Themes.cpp [hunk 6]
```diff
@@ -84,8 +88,9 @@ namespace OpenRCT2::Ui::Windows
...
-        MakeWidget({  5, 46}, {214,  15}, WindowWidgetType::TableHeader, WindowColour::Secondary, STR_THEMES_HEADER_WINDOW                                                           ), // Window header
-        MakeWidget({219, 46}, { 97,  15}, WindowWidgetType::TableHeader, WindowColour::Secondary, STR_THEMES_HEADER_PALETTE                                                          ), // Palette header
+        MakeWidget({  5, 46}, {kWindowHeaderWidth,  15}, WindowWidgetType::TableHeader, WindowColour::Secondary, STR_THEMES_HEADER_WINDOW                                                           ), // Window header
+        MakeWidget({157, 46}, { 79,  15}, WindowWidgetType::TableHeader, WindowColour::Secondary, STR_THEMES_HEADER_PALETTE                                                          ), // Palette header
+        MakeWidget({236, 46}, { 80,  15}, WindowWidgetType::TableHeader, WindowColour::Secondary, STR_THEMES_HEADER_TRANSPARENCY                                                          ), // Transparency header
```
[the remark this answers is not in the document]
Comment (lewyche) [reply src/openrct2-ui/windows/Themes.cpp:93]: Would this be better?
image

[the remark this answers is not in the document]
Comment (tupaschoal) [reply src/openrct2-ui/windows/Themes.cpp:93]: We typically let the line become long, but make sure the commas of all subsequent entries are aligned. That's real nitpick though, let us go through other concerns on the PR first.

File: src/openrct2-ui/windows/Themes.cpp [hunk 7]
```diff
@@ -244,12 +249,16 @@ namespace OpenRCT2::Ui::Windows
     {
     private:
         uint8_t _selected_tab = 0;
+        // index of class within tab
         int16_t _colour_index_1 = -1;
+        // index of button within class
         int8_t _colour_index_2 = -1;
```
[the remark this answers is not in the document]
Comment (lewyche) [reply src/openrct2-ui/windows/Themes.cpp:255]: I'm thinking something like `_class_index` and `_button_index`. Of course, I'll have to rename `GetColourIndex()` as well.

[the remark this answers is not in the document]
Comment (AaronVanGeffen) [reply src/openrct2-ui/windows/Themes.cpp:255]: Sounds good, though please use lowerCamelCase for variables. Here, that would mean `_classIndex`, `_buttonIndex`.

File: src/openrct2-ui/windows/Themes.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (lewyche) [reply src/openrct2-ui/windows/Themes.cpp:93]: Do you mean something like this?
image

**Comment by lewyche** [comment]
The rebase is broken right now, but I have to go. I will fix it later.

File: src/openrct2-ui/windows/Themes.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (tupaschoal) [reply src/openrct2-ui/windows/Themes.cpp]: Almost, I fixed it on my commit.

**Review by tupaschoal — APPROVED** [review]
I tested this for a bit and it seems to be working fine. Also fixed the formatting I mentioned and the changelog entry

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**